retention/test/test_cohort_analysis.py

```diff
@@ -396,9 +426,47 @@ def test_generate_by_bucket(self):
             (CohortKey(cohort_start_date='2018-10-05',
              client_type='web', sso_idp=''), 2, 1, None),
             (CohortKey(cohort_start_date='2018-10-05',
-             client_type='combined', sso_idp=''), 2, 3, 3)
+             client_type='combined', sso_idp=''), 2, 3, 3),
+            (CohortKey(cohort_start_date='2018-10-05',
+             client_type='iosx', sso_idp=''), 2, 0, None),
+            (CohortKey(cohort_start_date='2018-10-05',
+                       client_type='androidx', sso_idp=''), 2, 0, None),
         ])
 
+    def test_user_agent_to_client(self):
+        test_ua = [
+            # Electron
+            ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Element/1.11.33 Chrome/112.0.5615.183 Electron/24.3.1 Safari/537.36",
+             cohort_analysis.ELEMENT_ELECTRON),
+            (" Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Element/1.11.31 Chrome/112.0.5615.87 Electron/24.1.2 Safari/537.36",
+             cohort_analysis.ELEMENT_ELECTRON),
+
+            # Web
+            ("Mozilla/5.0 (Windows NT 10.0; rv:102.0) Gecko/20100101 Firefox/102.0",
+             cohort_analysis.WEB),
+            ("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36", cohort_analysis.WEB),
+
+            # Android
+            ("Element/1.5.25 (HUAWEI MOA-LX9N; Android 10; MOA-L49 3.1.0.110(C10E3R3P1); Flavour FDroid; MatrixAndroidSdk2 1.5.25)",
+                cohort_analysis.ELEMENT_ANDROID),
+            ("Element/1.5.32 (Google Pixel 6 Pro; Android 13; TQ2A.230505.002; Flavour GooglePlay; MatrixAndroidSdk2 1.5.32)",
+                cohort_analysis.ELEMENT_ANDROID),
+
+            # iOS
+            ("Element/1.10.12 (iPhone 14 Pro Max; iOS 16.4.1; Scale/3.00)",
+                cohort_analysis.ELEMENT_IOS),
+            ("Element/1.10.12 (iPhone 7 Plus; iOS 15.7.3; Scale/1.00)",
+                cohort_analysis.ELEMENT_IOS),
+
+            # Element X Android - TODO replace with real UA
```

Comment:
Here is what my Element X Android build sends as its user agent, captured from my Synapse logs:

```
Element X/0.1.0 (Google Pixel 4a (5G); Android 13; TQ3A.230605.011.2023062800; Sdk TODO)
```